### DISCUSSION TITLE
Литеральные множества [1,2,3]

### DISCUSSION THREAD

**Discussion by miks1965** [pascalabcnet/pascalabcnet#2421] [General]
Идея - сделать как в последнем Free Pascal - тип [1,2,3] определяется по контексту:

```pascal
procedure p(a: array of integer); begin end;
procedure q(s: set of integer); begin end;

begin
  p([1,2,3]);
  q([1,2,3]);
end.
```

С этой идеей не всё гладко. 

Проблема 1. Дело в том, что встроенные множества имеют также синтаксис [2..5,7]. И что с этим делать?

Есть 2 решения:

1. [2..5,7] считать всегда множеством, а [2,5,7] - по контексту
2. [2..5,7] по контексту преобразовывать в [2,3,4,5,7] - т.е. от чего уходим во множествах, к тому и возвращаемся. 

Что более ожидаемо?

Проблема 2. В записи [1,3,5] * [2,4,6], [1,3,5] + [2,4,6] чем считать литералы - множествами или массивами?
1. Считать множествами. Плюсы - совместимость со старым кодом. 
2. Считать массивами. Тогда первая запись не откомпилируется, а вторая вернёт конкатенированный массив. Ещё хуже - вот эта запись: [1,2,3] + [3..6] - это что будет? Ошибка компиляции? Последовательность?
3. Сделать ключ компиляции и переключать его. По умолчанию остановиться на 2.

Заметим, что в подавляющем большинстве случаев безразлично, как трактовать конструкцию [1,2,3]. Например, в выражении 2 in [1,2,3]

Конечно сильно беспокоит, как трактовать
```pascal
var a := [1,2,3];
```

Сейчас это множество, но хотелось бы массив.

Чтобы инициализировать можно было так:

```pascal
var a := [0] * 10;
var aa := [[1,2,3],[4,5]]
```

В некотором смысле простым решением будет считать и [2,4..6] массивом или множеством в зависимости от ключа компиляции. Всё равно множества неэффективны и кидаются в память - массивы в этом случае будут столь же неэффективны

**Comment by ibond84** [comment]
Зачем? Есть же |1,2,3|

**Comment by ibond84** [comment]
Тем более возникнут проблемы с перегрузками

```
procedure p(a: array of integer); begin end;
procedure p(s: set of integer); begin end;

```

**Reply by miks1965** [reply]
Можно сделать что [1,2,3] - массив, но массив неявно преобразуется к множеству

|1,2,3| - палочки плохо воспринимаются - [1,2,3] - более привычный синтаксис для массивов

**Comment by ibond84** [comment]
Я против

**Reply by miks1965** [reply]
Сейчас стал работать такой код: 
```
function operator implicit<T>(s: array of T): set of T; extensionmethod;
begin
  foreach var x in s do
    Include(Result,x)
end;

begin
  var s: set of integer := Arr(1,2,2);
  Print(s);
end.
```

То есть, массив можно неявно преобразовывать во множество.

Таким образом, запись [1,2,3,2] можно считать массивом. При преобразовании во множество вызовется implicit.

Если будут 2 перегрузки
```
procedure p(a: array of integer); begin end;
procedure p(s: set of integer); begin end;
``` 
то на `p([1,2,3,2])` выберется первая


Единственная проблема - это как трактовать конструкцию
```
[1,2,3,4] * [2,3,4,5]
```
В старом коде это было пересечение множеств, в новом по идее это будет ошибка компиляции.

Наверное, лучшим решением будет флаг, считать ли [1,2,3,2] множеством или массивом. И по умолчанию сделать массивом.

**Reply by spectatorBH** [reply]
Может тогда заодно по этому же флагу инвертировать и значение прямых скобок: т.е. сделать по дефолту `|1,2,3|` - множество, `[1,2,3]` - массив? А при наличии директивы/флага (напр., `$old_syntax`) -- наоборот. Тогда и пересечение множеств в простом виде для детей можно записывать, и с векторами/массивами нагляднее работать, и старый код не поломается. Не думаю, что недавно появившийся синтаксис прямых скобок широко используется, поэтому такая инверсия не должна вызвать больших проблем с новым поведением по умолчанию.

Под эту же директиву (точнее, ее отсутствие) можно было бы добавить в будущем и использование нормальных двойных кавычек для строковых литералов вместо неудобных одинарных (char не трогать) -- в англ. яз. и в ряде других европейских одинарная кавычка на клавиатуре совпадает с часто используемым в грамматике символом апосторофа, что сильно усложняет набор таких строк и эстетически их уродует. К тому же, это улучшит переносимость кода с/на C#.

Кроме того, так будет проще различать строки из одного символа от значений типа char (без использования символа интерполяции не по назначению). Также в новом синтаксисе можно бы ввести поддержку многострочных текстовых литералов (без добавления везде уродства в виде `#10#13` или `+ NewLine +`).

**Comment by miks1965** [comment]
Синтаксис |1,2,3|  уже активно используется. Я бы не менял местами. Я бы оставил  |1,2,3| для совместимости и сделал [1,2,3] для массивов в новом синтаксисе. Точнее, если не указан тип.

Скажем, в коде 
```
var s: set of integer := [1,2,2,3]
```
запись  [1,2,2,3] будет трактоваться как множество.

Другое предложение было трактовать [1,2,2,3] как List<integer>, а палочки оставить для массивов.
Учитывая автовыведение типа, в 90 процентах случаев никто не заметит разницу между списком и массивом.

Отдельная магия - в трактовке конструкции [].

К сожалению, проблема реализации оказалась сложнее чем мы думали изначально. 

Родственная проблема - желание реализовать list comprehensions раз уж есть синтаксис []

Здесь тоже много предложений:

```
[x*x for x in a if x >0] в стиле Питона
[for x in a where x > 0 select x*x] - в стиле LINQ
[x*x: x in a where x>0] - в математическом стиле


```

**Comment by miks1965** [comment]
Это уже сделано
